sql/src/main/scala/org/apache/spark/sql/geosparksql/expressions/Functions.scala

```diff
@@ -215,13 +216,27 @@ case class ST_Intersection(inputExpressions: Seq[Expression])
   extends Expression with CodegenFallback {
   override def nullable: Boolean = false
 
-  override def eval(input: InternalRow): Any = {
+
+  override def eval(inputRow: InternalRow): Any = {
     assert(inputExpressions.length == 2)
-    val leftgeometry = GeometrySerializer.deserialize(inputExpressions(0).eval(input).asInstanceOf[ArrayData])
-    val rightgeometry = GeometrySerializer.deserialize(inputExpressions(1).eval(input).asInstanceOf[ArrayData])
-    new GenericArrayData(GeometrySerializer.serialize(leftgeometry.intersection(rightgeometry)))
-  }
+    val leftgeometry = GeometrySerializer.deserialize(inputExpressions(0).eval(inputRow).asInstanceOf[ArrayData])
+    val rightgeometry = GeometrySerializer.deserialize(inputExpressions(1).eval(inputRow).asInstanceOf[ArrayData])
+
+    val isIntersects = leftgeometry.intersects(rightgeometry)
+    val isLeftContainsRight = leftgeometry.contains(rightgeometry)
+    val isRightContainsLeft = rightgeometry.contains(leftgeometry)
 
+
+    if (isIntersects && isLeftContainsRight) {
+      return new GenericArrayData(rightgeometry)
+    }
+
+    if (isIntersects && isRightContainsLeft) {
+      return new GenericArrayData(leftgeometry)
+    }
+
+    return new GenericArrayData(GeometrySerializer.serialize(leftgeometry.intersection(rightgeometry)))
```

Comment:
Thank you very much for implementing this!

But I think you forgot one case:

If (!isIntersects) return empty geometry

Reply:
Yes, I think you can do that. Just make sure you add one test case for this scenario.